packages/ui-components/src/components/QrCode/QrCodeVisual.tsx

```diff
@@ -0,0 +1,203 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+import React, { useState } from 'react';
+import styled from 'styled-components';
+import {
+  Button,
+  Checkbox,
+  DialogActions,
+  FormControl,
+  FormControlLabel,
+  FormGroup,
+  Typography,
+  Grid,
+  Container
+} from '@material-ui/core';
+import CheckboxIcon from '@material-ui/icons/CheckBox';
+import DownloadIcon from '@material-ui/icons/GetApp';
+import { NoData } from '../NoData';
+import { QrCodeImage } from './QrCodeImage';
+import { getCanvasUrlForDownload } from './useQrCodeCanvas'
+
+const FormContainer = styled.div`
+  display: flex;
+  justify-content: center;
+  padding: 0px 40px 30px 40px;
+`;
+
+const Error = styled.div`
+  color: ${props => props.theme.palette.error.main};
+  margin-top: 10px;
+  text-align: center;
+`;
+
+const StyledCanvas = styled.canvas`
+  outline: 1px solid #dedede;
+  width: 500px;
+  height: 500px;
+`;
+
+interface DownloadQrCodeVisualProps {
+  downloadImages: (qrCodeCanvasUrlsWithFileNames: {name: string, url: string | undefined}[]) => Promise<void>;
+  config?: object;
+  data?: { name: string; value: string }[];
+  isLoading?: boolean;
+  isEnlarged?: boolean;
+  onClose: () => void
+  className?: string;
+  error?: string;
+}
+
+export const QrCodeVisual = ({
+  downloadImages,
+  config = {},
+  data: options = [],
+  isLoading,
+  isEnlarged,
+  onClose,
+  className,
+  error,
+}: DownloadQrCodeVisualProps) => {
+
+  const noneSelected = Object.fromEntries(
+    options.map(({ value }) => [value, false]),
+  );
+  const [selectedQrCodes, setSelectedQrCodes] = useState(noneSelected);
+
+  const toggleSelectFile = (value: string) =>
+  setSelectedQrCodes({ ...selectedQrCodes, [value]: !selectedQrCodes[value] });
+
+  const [isDownloading, setIsDownloading] = useState(false);
```

Comment:
This is really pedantic, but can we put all the `useState` things at the top of the component, grouped together?